### PR TITLE
[qt] add mwms borders selection

### PR DESCRIPTION
**Pull request by gmoryes**
image
image

### CODE REVIEW

File: map/framework.cpp [hunk 1]
```diff
@@ -3496,13 +3496,59 @@ void VisualizeFeatureInRect(m2::RectD const & rect, FeatureType & ft, df::DrapeA
 }
 }  // namespace
 
+void VisualizeMwmBorder(df::DrapeApi & drapeApi, std::string const & mwmName)
```
Comment (mpimenov) [inline map/framework.cpp]: Can we have a mode that either chooses the .poly file or unpacks the actual polygon from packed_polygons.bin?

File: map/framework.cpp [hunk 2]
```diff
@@ -3496,13 +3496,59 @@ void VisualizeFeatureInRect(m2::RectD const & rect, FeatureType & ft, df::DrapeA
...
+    if (!strings::to_double(*iter, lon))
+      return;
+    ++iter;
+
+    if (!strings::to_double(*iter, lat))
+      return;
+
+    points.emplace_back(MercatorBounds::FromLatLon(lat, lon));
+  }
+
```
Comment (darina) [inline map/framework.cpp]:
```
if (points.empty())
  return;
```
или CHECK

Comment (gmoryes) [reply map/framework.cpp]: там может прийти пустой массив, в конце файла data/borders/*.poly 
пишется 
```sh
END
END
```
после каждого END будем пытаться рисовать

Comment (gmoryes) [reply map/framework.cpp]:
```cpp
if (points.empty())
  return;
```
done

**Comment by gmoryes** [comment]
при нажатии на кнопку теперь выскакивает такая менюшка
image

**Comment by gmoryes** [comment]
@mpimenov PTAL

File: qt/draw_widget.cpp [hunk 3]
```diff
@@ -35,6 +38,7 @@
 #include <QtCore/QLocale>
 #include <QtCore/QThread>
 #include <QtCore/QTimer>
+#include <base/file_name_utils.hpp>
```
Comment (mpimenov) [inline qt/draw_widget.cpp]: ?

File: qt/draw_widget.cpp [hunk 4]
```diff
@@ -208,7 +212,7 @@ void DrawWidget::mousePressEvent(QMouseEvent * e)
     {
       SubmitBookmark(pt);
     }
-    else if (!(m_selectionMode || m_cityBoundariesSelectionMode || m_cityRoadsSelectionMode) ||
+    else if (!(m_currentSelectionMode) ||
```
Comment (mpimenov) [inline qt/draw_widget.cpp]: Please drop the parentheses

File: qt/draw_widget.cpp [hunk 5]
```diff
@@ -233,13 +237,30 @@ void DrawWidget::mouseMoveEvent(QMouseEvent * e)
...
       m_rubberBand != nullptr && m_rubberBand->isVisible())
   {
     m_rubberBand->setGeometry(QRect(m_rubberBandOrigin, e->pos()).normalized());
   }
 }
 
+void DrawWidget::VisualizeMwmsBordersInRect(m2::RectD const & rect, bool withPoints)
+{
+  static std::string const kPathToBorders =
+    base::JoinPath(GetPlatform().ResourcesDir(), "borders");
```
Comment (mpimenov) [inline qt/draw_widget.cpp]: Can we have a mode to visualize the borders from packed_polygons.bin?

File: qt/draw_widget.cpp [hunk 6]
```diff
@@ -233,13 +237,30 @@ void DrawWidget::mouseMoveEvent(QMouseEvent * e)
...
+
+  auto const mwmNames = m_framework.GetRegionsCountryIdByRect(rect, false /* rough */);
+
+  for (auto const & mwmName : mwmNames)
+  {
+    std::string const path = base::JoinPath(kPathToBorders, mwmName + ".poly");
```
Comment (mpimenov) [inline qt/draw_widget.cpp]: BORDERS_EXTENSION?

File: qt/draw_widget.cpp (hunk 5 above)
Comment (mpimenov) [inline qt/draw_widget.cpp]: BORDERS_DIR?

File: qt/mwms_borders_selection.hpp [hunk 7]
```diff
@@ -0,0 +1,36 @@
+#pragma once
+
+#include <QtWidgets/QApplication>
```
Comment (mpimenov) [inline qt/mwms_borders_selection.hpp]: qt after stdlib

File: qt/mwms_borders_selection.hpp [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+{
+public:
+  MwmsBordersSelection(QWidget * parent);
+
+  enum class Response
+  {
+    JustBorders,
+    WithPointsAndBorders,
+
+    Canceled
```
Comment (mpimenov) [inline qt/mwms_borders_selection.hpp]: s/Canceled/Cancelled/

File: qt/draw_widget.cpp (hunk 5 above)
Comment (gmoryes) [reply qt/draw_widget.cpp]: image

Comment (gmoryes) [reply qt/draw_widget.cpp]: image
Moscow and Moscow Oblast_West with borders from packed_polygon.bin

**Comment by gmoryes** [comment]
@mpimenov PTAL

**Comment by gmoryes** [comment]
JTA

**Comment by gmoryes** [comment]
JTMAC

File: qt/draw_widget.cpp (hunk 5 above)
Comment (mpimenov) [reply qt/draw_widget.cpp]: Please also add an option to show the bounding box.

File: map/framework.cpp [hunk 9]
```diff
@@ -3497,13 +3499,44 @@ void VisualizeFeatureInRect(m2::RectD const & rect, FeatureType & ft, df::DrapeA
 }
 }  // namespace
 
+void Framework::DrawMwmBorder(std::string const & mwmName,
+                              std::vector<m2::RegionD> const & polygons, bool withPoints)
```
Comment (mpimenov) [inline map/framework.cpp]: s/polygons/regions/ ?

Comment (mpimenov) [inline map/framework.cpp]: s/withPoints/withVertices/ ?

File: map/framework.cpp [hunk 10]
```diff
@@ -3497,13 +3499,44 @@ void VisualizeFeatureInRect(m2::RectD const & rect, FeatureType & ft, df::DrapeA
...
+    if (points.empty())
+      return;
+
+    static uint32_t kColorCounter = 0;
+    if (withPoints)
+    {
+      m_drapeApi.AddLine(mwmName, df::DrapeApiLineData(points, colorList[kColorCounter])
+          .Width(4.0f)
+          .ShowId()
+          .ShowPoints(true));
```
Comment (mpimenov) [inline map/framework.cpp]: true /* markPoints */

File: map/framework.cpp [hunk 11]
```diff
@@ -3497,13 +3499,44 @@ void VisualizeFeatureInRect(m2::RectD const & rect, FeatureType & ft, df::DrapeA
...
+    }
+    else
+    {
+      m_drapeApi.AddLine(
+          mwmName, df::DrapeApiLineData(points, colorList[kColorCounter]).Width(4.0f).ShowId());
```
Comment (mpimenov) [inline map/framework.cpp]:
```
auto lineData = df::DrapeApiLineData(points, colorList[kColorCounter]).Width(4.0f).ShowId();
if (withPoints)
  lineData.ShowPoints(true /* markPoints */);
m_drapeApi.AddLine(mwmName, lineData);
```

File: qt/draw_widget.cpp (hunk 5 above)
Comment (gmoryes) [reply qt/draw_widget.cpp]: image

**Comment by gmoryes** [comment]
@mpimenov PTAL

**Comment by gmoryes** [comment]
@mpimenov PTAL

**Pull request merged**